### PR TITLE
kie-issues#174: Support constrained Data Types on inputs table of DMN Runner

### LINKED ISSUE #174: Support constrained Data Types on DMN Runner form
Adding a "Range" or "Expression" constrait for some Data Types add a "x-dmn-allowed-values" property to the model JSON Schema. This property is not used in the validation, enabling the user to add any values in the DMN Runner form without a warning. Adding a constraint to a "date", "time", "date-time", "days and time duratation", "years and months duration" will add the "x-dmn-allowed-values".

```[tasklist]
### Acceptance criteria
- [ ] Adding a "Range" constraint to a Data Type should reflect on DMN Runner form validation.
- [ ] Adding a "Expression" constraint to a Data Type should reflect on DMN Runner form validation.
```

### PR DESCRIPTION
**Pull request by jomarko**
Closes kiegroup/kie-issues#174

This PR:
- Set correct column minimal width for custom DMN data types that are date or time based, e.g. `myDateType: date`
- Check if there is `Range` constraint for dmn custom type and validates values put into DMN runner form [1]
- Check if there is `Enumeration` constraint for dmn custom type and validates values put into DMN runner form
- **does not** check free form **expression** constraints that can user put on DMN custom data types

[1]
Screenshot 2023-06-05 155431
Screenshot 2023-06-05 155422
Screenshot 2023-06-07 154347

### CODE REVIEW

**Review by ljmotta — COMMENTED** [review]
@jomarko  Thanks for taking care of this issue.

We shouldn't add any DMN Runner specific checks or/and  DMN logic inside the `uniforms-patternfly` library. Its sole purpose is to generate a form based on a JSON Schema, and any other thing not related to the JSON Schema spec or the form generation shouldn't be in this library.

Regarding the data validation, we are just showing the inline errors in the Form view, as it will overlap the table rows.

Now, to address this issue, a more appropriate approach would verify if it's possible to add this check in our Ajv setup in the `packages/dmn-runner/ajv.ts`. Take a look in the Ajv docs on how to add [keywords](https://ajv.js.org/keywords.html).

If it isn't possible to add the validation in the Ajv setup, we should change our form and table `uniforms` Bridge.
 - form: `packages/form-dmn/src/uniforms/FormDmnJsonSchemaBridge.ts`
 - table: `packages/unitables-dmn/src/uniforms/DmnUnitablesJsonSchemaBridge.ts`

**Comment by tiagobento** [comment]
@jomarko Formatting is broken for this PR... Can you please take a look?

File: packages/dmn-runner/src/ajv.ts [hunk 1]
```diff
@@ -39,7 +132,87 @@ export class DmnRunnerAjv {
...
+              if (minAllowedDate && compareMinBoundary(actualDate, minAllowedDate)) {
+                return false;
+              }
+              const maxAllowedDate = new Date(maxAllowed);
+              if (maxAllowedDate && compareMaxBoundary(actualDate, maxAllowedDate)) {
+                return false;
+              }
+            }
+          }
+          if (enumeratedValues) {
```
Comment (yesamer) [inline packages/dmn-runner/src/ajv.ts]: @jomarko A type can have an `allowedValues` only, I'd use a "else if" statement:

eg. 
`if (minAllowed && maxAllowed) {
 ....
} else if (enumeratedValues) {
 .....
`

**Review by yesamer — COMMENTED** [review]
Thank you @jomarko, great feature. 

In my local, Range validation (for `date`) doesn't work as expected, please check the above video. It's weird because I guess you already tested that 

https://github.com/kiegroup/kie-tools/assets/16005046/08c181ad-4b46-4413-b440-3f08ae9228b8

File: packages/dmn-runner/src/ajv.ts [hunk 2]
```diff
@@ -39,7 +132,87 @@ export class DmnRunnerAjv {
...
+
+        return (data: string) => {
+          if (data.includes(".") && data.endsWith("Z")) {
+            // adjusting from "2023-06-01T10:42:00.000Z" to "2023-06-01T10:42:00"
+            data = data.substring(0, data.lastIndexOf("."));
+          }
+          if (minAllowed && maxAllowed) {
+            // It is a Range constraint
+            if (parentSchema.format === "time") {
+              if (compareMinBoundary(data, minAllowed ?? "") || compareMaxBoundary(data, maxAllowed ?? "")) {
```
Comment (yesamer) [inline packages/dmn-runner/src/ajv.ts]: `?? ""` are not necessary, based on the `if` in line 163

```suggestion
              if (compareMinBoundary(data, minAllowed) || compareMaxBoundary(data, maxAllowed)) {
```

File: packages/dmn-runner/src/ajv.ts [hunk 3]
```diff
@@ -39,7 +132,87 @@ export class DmnRunnerAjv {
...
+          parentSchema.format !== "years and months duration"
+        ) {
+          return (data: string) => true;
+        }
+        const { minAllowed, minAllowedIncluded, maxAllowed, maxAllowedIncluded } = this.parseRangeFromAllowedValues(
+          schema ?? "",
+          parentSchema.format
+        );
+        const enumeratedValues = this.parseEnumerationFromAllowedValues(schema ?? "", parentSchema.format);
+        const compareMinBoundary: (a: Date | String | number, b: Date | String | number) => boolean = minAllowedIncluded
```
Comment (yesamer) [inline packages/dmn-runner/src/ajv.ts]: I would rename that to `isUnderTheMinBoundary` to better clarify the scope.
I would rename `a` to `actual` or `value`, and `b` to `minBoundary`

File: packages/dmn-runner/src/ajv.ts (hunk 1 above)
Comment (jomarko) [reply packages/dmn-runner/src/ajv.ts]: will do, thank you, good suggestion

File: packages/dmn-runner/src/ajv.ts (hunk 2 above)
Comment (jomarko) [reply packages/dmn-runner/src/ajv.ts]: well spotted, removing

File: packages/dmn-runner/src/ajv.ts (hunk 3 above)
Comment (jomarko) [reply packages/dmn-runner/src/ajv.ts]: perfect suggestion for renaming, will do

File: packages/dmn-runner/src/ajv.ts [hunk 4]
```diff
@@ -29,6 +38,90 @@ import {
 export class DmnRunnerAjv {
   private ajv;
 
+  private parseRangeFromAllowedValues = (
+    xDmnAllowedValues: string,
+    units: "date" | "time" | "date-time" | "days and time duration" | "years and months duration"
```
Comment (ljmotta) [inline packages/dmn-runner/src/ajv.ts]: I don't think `units` correctly capture what this variable is. The method invocation passes  down a `type` in this parameter, so I guess `type` or `dmnType` would be better.

Also, you should create an enum in this case, to avoid repeated strings in your code.

```
enum DmnType {
  DATE = "date",
  TIME = "time",
  ...
}

File: packages/dmn-runner/src/ajv.ts [hunk 5]
```diff
@@ -29,6 +38,90 @@ import {
...
+  ) => {
+    if (xDmnAllowedValues) {
```
Comment (ljmotta) [inline packages/dmn-runner/src/ajv.ts]: The `xDmnAllowedValues` is a string, it can't be `undefined` or `null`, so this if statement is not necessary, unless you're checking for an empty string. I suggest to add a short circuit "if" to avoid the entire code to be inside of one big if: 

```ts
if (xDmnAllowedValues === "") {
  return {}
}
```

```suggestion
```

File: packages/dmn-runner/src/ajv.ts [hunk 6]
```diff
@@ -29,6 +38,90 @@ import {
...
+          const maxAllowed = matches[1].replace(`${feelFunction}("`, "").replace('")', "");
+          const maxAllowedIncluded = xDmnAllowedValues.endsWith("]");
+          return { minAllowed, minAllowedIncluded, maxAllowed, maxAllowedIncluded };
+        }
+      }
+    }
+    return {};
+  };
+
+  private parseEnumerationFromAllowedValues = (
```
Comment (ljmotta) [inline packages/dmn-runner/src/ajv.ts]: Same comments as the above method.

File: packages/dmn-runner/src/ajv.ts [hunk 7]
```diff
@@ -39,7 +132,87 @@ export class DmnRunnerAjv {
     });
     this.ajv.addMetaSchema(metaSchemaDraft04);
     this.ajv.addKeyword(X_DMN_TYPE, {});
-    this.ajv.addKeyword(X_DMN_ALLOWED_VALUES, {});
+    this.ajv.addKeyword(X_DMN_ALLOWED_VALUES, {
+      compile: (schema, parentSchema: { format?: string }, it) => {
```
Comment (ljmotta) [inline packages/dmn-runner/src/ajv.ts]: With your enum you could add a more specific type here.

File: packages/unitables/src/uniforms/UnitablesJsonSchemaBridge.tsx [hunk 8]
```diff
@@ -158,6 +158,14 @@ export class UnitablesJsonSchemaBridge extends JSONSchemaBridge {
             type: field.type,
           };
         }
+        switch (field["format"]) {
```
Comment (ljmotta) [inline packages/unitables/src/uniforms/UnitablesJsonSchemaBridge.tsx]: Please, use the `field.format` way. The `[" ... "]` should be used only when strictly necessary. Also, the format check could be moved to inside the line 72 if statement to reuse the switch case.

**Review by ljmotta — CHANGES_REQUESTED** [review]
@jomarko Hi! The PR looks great. I've made some specific comments inline.

File: packages/dmn-runner/src/ajv.ts (hunk 5 above)
Comment (jomarko) [reply packages/dmn-runner/src/ajv.ts]: good point, thank you, removing

File: packages/dmn-runner/src/ajv.ts (hunk 6 above)
Comment (jomarko) [reply packages/dmn-runner/src/ajv.ts]: thank you, removing also here

File: packages/unitables/src/uniforms/UnitablesJsonSchemaBridge.tsx (hunk 8 above)
Comment (jomarko) [reply packages/unitables/src/uniforms/UnitablesJsonSchemaBridge.tsx]: thank you @ljmotta I understand `"format"` vs `.format`, but I am not sure how should look code once I move it to line 72. could you please share some pseudocode at least?

File: packages/dmn-runner/src/ajv.ts (hunk 7 above)
Comment (jomarko) [reply packages/dmn-runner/src/ajv.ts]: good point

File: packages/dmn-runner/src/ajv.ts (hunk 4 above)
Comment (jomarko) [reply packages/dmn-runner/src/ajv.ts]: good point, introducing one enum

File: packages/unitables/src/uniforms/UnitablesJsonSchemaBridge.tsx (hunk 8 above)
Comment (ljmotta) [reply packages/unitables/src/uniforms/UnitablesJsonSchemaBridge.tsx]: I didn't test it to check some corner cases. But what I meant was to make this change on line 72:

```ts
if (!xDmnType) {
  if (field.format) {
    type = field.format
  } else {
    type = field.type;
  }
} 
```

Then you will not need this new swtich case.

Comment (jomarko) [reply packages/unitables/src/uniforms/UnitablesJsonSchemaBridge.tsx]: Understand now, will do, thank you!

Comment (jomarko) [reply packages/unitables/src/uniforms/UnitablesJsonSchemaBridge.tsx]: @ljmotta I pushed the changes in a separate commit [8b49916](https://github.com/kiegroup/kie-tools/pull/1693/commits/8b49916c0022dd0db32dcb3bb8b791275964ba7f)

However I needed to put a code to slightly different place, because it is not true `!xDmnType` is not set. See screenshot from my debuging.

Screenshot 2023-06-13 115859

**Comment by jomarko** [comment]
@yesamer I struggle to reproduce the issue from your video. I changed my browser locale and system setting to 'Slovak', to be totally different than english,

Snímka obrazovky 2023-06-13 140523


However #1693 seems to work properly on my machine. May I ask you to re-check once more on your machine?

@ljmotta all points should be addressed now.

**Comment by yesamer** [comment]
@jomarko Now it works well, probably one of the changes you applied "fixed" that.

**Review by yesamer — APPROVED** [review]

**Comment by ljmotta** [comment]
Hi @jomarko I was performing some manual checks, and I found a bug, please check the following scenario:

https://github.com/kiegroup/kie-tools/assets/24302289/40158252-6833-4532-9117-41b18a8570b4

I've already opened a PR to your branch solving the issue. https://github.com/jomarko/kie-tools/pull/8

Another thing that come to my attention is that the "validation" error and the calendar icon is overlaping, this is happening for you too or it's a problem in my side? I guess it's a Patternfly bug.

**Review by ljmotta — APPROVED** [review]
